v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+    /* Codes_SRS_LOG_SINK_ETW_01_004: [ If the state is NOT_REGISTERED (1): ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_005: [ log_sink_etw_log shall switch the state to REGISTERING (2). ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_009: [ Checking and changing the variable that maintains whether TraceLoggingRegister was called shall be done using InterlockedCompareExchange and InterlockedExchange. ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_011: [ If the state is REGISTERED (3), log_sink_etw_log shall proceed to log the ETW event. ]*/
+    /* Codes_SRS_LOG_SINK_ETW_01_087: [ If the state is REGISTERING (2) log_sink_etw_log shall wait until the state is not REGISTERING (2). ]*/
+    while ((state = InterlockedCompareExchange(&etw_provider_state, PROVIDER_STATE_REGISTERING, PROVIDER_STATE_NOT_REGISTERED)) != PROVIDER_STATE_REGISTERED)
+    {
+        if (state == PROVIDER_STATE_NOT_REGISTERED)
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_006: [ log_sink_etw_log shall register the ETW TraceLogging provider by calling TraceLoggingRegister (TraceLoggingRegister_EventRegister_EventSetInformation). ]*/
+            TLG_STATUS register_result = TraceLoggingRegister(g_my_component_provider);
+            if (SUCCEEDED(register_result))
+            {
+                /* Codes_SRS_LOG_SINK_ETW_01_007: [ log_sink_etw_log shall switch the state to REGISTERED. ]*/
+                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_REGISTERED);
+                /* Codes_SRS_LOG_SINK_ETW_01_008: [ log_sink_etw_log shall printf the fact that the provider was registered and from which executable (as obtained by calling _pgmptr). ]*/
+
+                char* executable_full_path_name;
+                if (_get_pgmptr(&executable_full_path_name) != 0)
+                {
+                    /* Codes_SRS_LOG_SINK_ETW_01_083: [ If _get_pgmptr fails, the executable shall be printed as UNKNOWN. ]*/
+                    executable_full_path_name = "UNKNOWN";
+                }
+
+                /* Codes_SRS_LOG_SINK_ETW_01_008: [ log_sink_etw_log shall emit a LOG_LEVEL_INFO event as a self test , printing the fact that the provider was registered and from which executable (as obtained by calling _get_pgmptr). ]*/
+                log_sink_etw_log(LOG_LEVEL_INFO, NULL, __FILE__, __FUNCTION__, __LINE__, "ETW provider was registered succesfully (self test). Executable file full path name = %s", MU_P_OR_NULL(executable_full_path_name));
+
+                break;
+            }
+            else
+            {
+                (void)printf("ETW provider was NOT registered.\r\n");
+                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_NOT_REGISTERED);
+            }
+        }
+    }
+}
+
+// want these aligned to 1 byte (copy/pasted from Tracelogging code)
+__pragma(pack(push, 1))
+typedef struct SELF_DESCRIBED_EVENT_TAG
+{
+    uint8_t _tlgBlobTyp;
+    uint8_t _tlgChannel;
+    uint8_t _tlgLevel;
+    uint8_t _tlgOpcode;
+    uint64_t _tlgKeyword;
```

Comment:
>_tlgKeyword [](http://example.com/codeflow?start=13&length=11)

Oh uh, who takes care of writing this?

1) the compiler does it because it is magic.
2) we do it byte by byte (cause I don't think this is aligned at this moment).

Reply:
compiler does it.
Raymond has a nice block where he calls us RISC haters. Which we are. Well, Tracelogging is too I guess.

[https://devblogs.microsoft.com/oldnewthing/20200103-00/?p=103290](https://devblogs.microsoft.com/oldnewthing/20200103-00/?p=103290)